File: datasets/ICDAR/rect.py
class Rect(object):
  def __init__(self, left, top, right, bottom):
    self.left = left
    self.top = top
    self.right = right
    self.bottom = bottom

  def as_tuple(self):
    return (self.left, self.top, self.right, self.bottom)

  def contains(self, other):
    return (
      self.left <= other.left and other.right <= self.right
      and self.top <= other.top and other.bottom <= self.bottom
    )

  def __eq__(self, other):
    return self.as_tuple() == other.as_tuple()

File: ops/rect.cpp
#include "rect.h"
#include <cassert>
#include <algorithm>

using std::min;
using std::max;


Rect::Rect(int left, int top, int right, int bottom) :
    Left(left),
    Top(top),
    Right(right),
    Bottom(bottom)
{
    assert(Left <= Right && Top <= Bottom);
}

bool Rect::IsEmpty() const
{
    return Left == Right || Top == Bottom;
}

bool Rect::HasIntersection(const Rect& other) const
{
    return Left < other.Right && other.Left < Right
        && Top < other.Bottom && other.Top < Bottom;
}

Rect Rect::operator|(const Rect& other) const
{
    const int left = min(Left, other.Left);
    const int top = min(Top, other.Top);
    const int right = max(Right, other.Right);
    const int bottom = max(Bottom, other.Bottom);
    return Rect(left, top, right, bottom);
}